File: 143_Reorder_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if (head == NULL || head->next == NULL || head->next->next == NULL) return;
        ListNode *p1 = head;
        ListNode *p2 = head;

        // Find the middle point
        while (p2 && p2->next) {
            p1 = p1->next;
            p2 = p2->next->next;
        }

        // Reverse the latter part of the linked list
        ListNode *head2 = p1->next;
        p1->next = NULL;
        p2 = head2->next;
        head2->next = NULL;
        while (p2) {
            p1 = p2->next;
            p2->next = head2;
            head2 = p2;
            p2 = p1;
        }

        // Reorder the list
        p1 = head;
        p2 = head2;
        while (p2) {
            auto temp = p1->next;
            p1 = p1->next = p2;
            p2 = temp;
        }
    }
};

File: 143_Reorder_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if head is None or head.next is None:
            return
        slow, fast = head, head.next
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
        tail = slow.next
        curr = tail.next
        slow.next = None
        tail.next = None
        while curr is not None:
            temp = curr
            curr = curr.next
            temp.next = tail
            tail = temp
        
        curr = head
        while curr is not None and tail is not None:
            temp = curr.next
            curr.next = tail
            curr = temp
            temp = tail.next
            tail.next = curr
            tail = temp
